src/main/java/com/github/maleksandrowicz93/cqrsdemo/student/StudentController.java

```diff
@@ -0,0 +1,82 @@
+package com.github.maleksandrowicz93.cqrsdemo.student;
+
+import com.github.maleksandrowicz93.cqrsdemo.student.api.StudentApi;
+import com.github.maleksandrowicz93.cqrsdemo.student.dto.SaveStudentRequest;
+import com.github.maleksandrowicz93.cqrsdemo.student.dto.StudentDto;
+import com.github.maleksandrowicz93.cqrsdemo.student.dto.StudentIdentification;
+import lombok.RequiredArgsConstructor;
+import lombok.experimental.FieldDefaults;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.servlet.support.ServletUriComponentsBuilder;
+
+import java.net.URI;
+import java.util.List;
+import java.util.UUID;
+
+@RestController
+@FieldDefaults(makeFinal = true)
+@RequiredArgsConstructor
+class StudentController implements StudentApi {
+
+    StudentFacade studentFacade;
+
+    @Override
+    public ResponseEntity<List<StudentIdentification>> findAllStudents(Integer page, Integer size) {
+        List<StudentIdentification> students = studentFacade.getAllStudents(page, size);
+        return ResponseEntity.ok(students);
+    }
+
+    @Override
+    public ResponseEntity<StudentDto> addStudent(SaveStudentRequest saveStudentRequest) {
+        return studentFacade.addStudent(saveStudentRequest)
+                .map(student -> {
+                    String location = getLocation(student.id());
+                    return ResponseEntity
+                            .created(URI.create(location))
+                            .body(student);
+                })
+                .orElseGet(() -> {
+                    var id = studentFacade.findStudentIdByEmail(saveStudentRequest.email());
```

Comment:
Ta informacja mogłaby być już zwracana w rezultacje `addStudent`

Reply:
Nie podoba mi się takie API, które jeśli wszystko jest ok, to zwraca normalnie wypełnione StudentDto(które może mieć wypełnione na dobrą sprawę tylko id i email), a jeśli student nie może być dodany, to zwraca StudentDto wypełnione częściowo (tylko id). Zmusza to klienta API do interpretacji wyniku poprzez analizę które pole jest wypełnione, a które nie. Jakakolwiek tutaj zmiana logiki będzie implikowała zmianę sposobu analizy pól w StudentDto. Dlatego właśnie tak do tego podszedłem, żeb interpretacja była jedynie na zasadnie czy jest coś w Optionallu, czy nie. Jeśli miałbym to jakoś uprościć, to przychodzi mi do głowy takie rozwiązanie:
if(facade.metodaAnalizującaStan) {
    return ResponseEntity jeśli stan nie ok
}
facade.addStudent
return ResponseEntity.ok

Reply:
Wystarczy stworzyć klasę obsługującą rezultat i tam możesz zawrzeć wszystko, bez potrzeby opierania się o binarny rezultat typu Optional, skoro w tym przypadku wynik metody to jedna z N opcji, a nie jedna z dwóch.

Reply:
Kumam - stworzyć jakiś customowy wrapper, który posiada więcej informacji niż Optional (jest, nie ma) i tam wrzucić wszystkie informacje i wtedy easy sobie obsłużę wynik w kontrolerze. Wtedy też dużo łatwiej będzie pozbyć się wyjątków. Za bardzo w głowie mam zakorzeniony styl: model/optional + exceptiony (na każdym projekcie architektura cebuli). Ale powoli wykorzeniamy tą cebulę 😄